homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,282 @@
+        standard = ClientResponseError(
+            response.request_info,
+            response.history,
+            code=response.status,
+            headers=response.headers,
+        )
+        data = await response.text()
+        _LOGGER.error("Request failed: %s", data)
+        raise InvalidAuthError(data) from standard
+
+
+WANTED_SCOPES = {"openid", "email", "profile"}
+
+
+@AUTH_PROVIDERS.register("openid")
+class OpenIdAuthProvider(AuthProvider):
+    """Example auth provider based on hardcoded usernames and passwords."""
+
+    DEFAULT_TITLE = "OpenID Connect"
+
+    _configuration: Dict[str, Any]
+    _jwks: Dict[str, Any]
+
+    async def async_get_configuration(self) -> Dict[str, Any]:
+        """Get discovery document for OpenID."""
+        session = async_get_clientsession(self.hass)
+        async with session.get(self.config[CONF_CONFIGURATION]) as response:
+            await raise_for_status(response)
+            data = await response.json()
+        return cast(Dict[str, Any], OPENID_CONFIGURATION_SCHEMA(data))
+
+    async def async_get_jwks(self) -> Dict[str, Any]:
+        """Get the keys for id verification."""
+        session = async_get_clientsession(self.hass)
+        async with session.get(self._configuration["jwks_uri"]) as response:
+            await raise_for_status(response)
+            data = await response.json()
+        return cast(Dict[str, Any], data)
+
+    @property
+    def redirect_uri(self) -> str:
+        """Return the redirect uri."""
+        url = URL(get_url(self.hass, prefer_external=True, allow_cloud=False))
+        return str(url.with_path(AUTH_CALLBACK_PATH))
+
+    async def async_login_flow(self, context: Optional[Dict]) -> LoginFlow:
+        """Return a flow to login."""
+
+        if not hasattr(self, "_configuration"):
+            self._configuration = await self.async_get_configuration()
+
+        if not hasattr(self, "_jwks"):
+            self._jwks = await self.async_get_jwks()
+
+        async_register_view(self.hass)
+
+        return OpenIdLoginFlow(self)
+
+    async def async_retrieve_token(self, code: str) -> Dict[str, Any]:
+        """Convert a token code into an actual token."""
+
+        payload = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": self.redirect_uri,
+            "client_id": self.config[CONF_CLIENT_ID],
+            "client_secret": self.config[CONF_CLIENT_SECRET],
+        }
+
+        session = async_get_clientsession(self.hass)
+        async with session.post(
+            self._configuration["token_endpoint"], data=payload
+        ) as response:
+            await raise_for_status(response)
+            return cast(Dict[str, Any], await response.json())
+
+    async def async_validate_token(
+        self, token: Dict[str, Any], nonce: str
+    ) -> Dict[str, Any]:
+        """Validate a token."""
+        from jose import jwt  # noqa: pylint: disable=import-outside-toplevel
```

Comment:
Why not import at the top?

Reply:
Oh I guess `REQUIREMENTS` is used and not a manifest.

Reply:
Why i wanted to get this into the normal integration logic :)